values.yaml

```diff
@@ -1,402 +1,160 @@
-  # be left blank, and the user can still access the platform from the external
-  # IP or a DNS name that resolves to the external IP address.
-  host: ""
-  # ingress.loadBalancerIP -- sets the external IP address of the ingress to the
-  # provided value.
-  loadBalancerIP: ""
-  # ingress.loadBalancerSourceRanges -- If specified and supported by the
-  # platform, the traffic through the cloud-provider load-balancer will be
-  # restricted to the specified client IPs. This field will be ignored if the
-  # cloud-provider does not support the feature.
-  loadBalancerSourceRanges: []
-  # ingress.podSecurityPolicyName -- The name of the pod security policy the
-  # built-in ingress controller should abide. It should be noted that the
-  # ingress controller requires the `NET_BIND_SERVICE` capability, privilege
-  # escalation, and access to privileged ports to successfully deploy. Ignored
-  # if `ingress.useDefault` is false.
-  podSecurityPolicyName: ""
-  # ingress.additionalAnnotations -- Deprecated. Please use `ingress.annotations`.
-  additionalAnnotations: []
-  # ingress.annotations -- Additional annotations to be used when creating the
-  # ingress. These only apply to the Ingress Kubernetes kind. The annotations
-  # can be used to specify certificate issuers or other cloud provider-specific
-  # integrations.
-  annotations: {}
-  # ingress.tls -- TLS options for the ingress. The hosts used for the tls
-  # configuration come from the ingress.host and the devurls.host variables. If
-  # those don't exist, then the TLS configuration will be ignored.
+# coderd -- Primary service responsible for all things Coder!
+coderd:
+  # coderd.image -- Injected by Coder during release.
+  image: ""
+  # coderd.replicas -- The number of Kubernetes Pod replicas.
+  replicas: 1
+  # coderd.serviceSpec -- Specification to inject for the coderd service. See:
+  # https://kubernetes.io/docs/concepts/services-networking/service/
+  serviceSpec:
+    # coderd.serviceSpec.type -- Set the type of Service. See:
+    # https://kubernetes.io/docs/concepts/services-networking/service/#publishing-services-service-types
+    type: LoadBalancer
+    # coderd.serviceSpec.externalTrafficPolicy -- Set the traffic policy for the service. See:
+    # https://kubernetes.io/docs/tasks/access-application-cluster/create-external-load-balancer/#preserving-the-client-source-ip
+    externalTrafficPolicy: Local
+    # coderd.serviceSpec.loadBalancerIP -- Set the external IP address of the
+    # Ingress service.
+    loadBalancerIP: ""
+    # coderd.serviceSpec.loadBalancerSourceRanges -- Traffic through the LoadBalancer
+    # will be restricted to the specified client IPs. This field will be ignored if
+    # the cloud provider does not support this feature.
+    loadBalancerSourceRanges: []
+  # coderd.tls -- TLS configuration for coderd.
+  # These options will override dashboard configuration.
   tls:
```

Comment:
@kylecarbs What do you think of this?

The goal is to make it easy for people to configure any number of certificates
and keys, which can be used for different purposes:

1. Server certificate to offer to users
2. CA certificate to validate external services
3. Client certificate to offer as authentication to external services (mTLS)

The default secretNames will be used as CA certs (scanned using the certificate
handling code, since we look recursively for these files)

```yaml
tls:
  certificates:
    # A list of secret names. The contents of this secret would be mounted
    # to /etc/ssl/certs with the same name (e.g., /etc/ssl/certs/ca-certs)
    secretNames:
      - ca-certs
      - other-certs
    # A list of files stored under ./ssl/certs. The contents of these files
    # would be added as a secret that Coder creates. We can use Files.Glob
    # to load the files into a Secret or ConfigMap: https://helm.sh/docs/chart_template_guide/accessing_files/#configmap-and-secrets-utility-functions
    # Maybe ConfigMap makes sense since certificates are public keys, not
    # really secret.
    files:
      - ca.crt
      - server.crt
  keys:
    # A list of secret names. The contents of this secret would be mounted
    # to /etc/ssl/private with the same name (e.g., /etc/ssl/private/ca-certs)
    secretNames:
      - server-keys
      - client-keys
    # A list of files stored under ./ssl/private. The contents of these files
    # would be added as a secret that Coder creates. We can use Files.Glob
    # to load the files into a Secret or ConfigMap: https://helm.sh/docs/chart_template_guide/accessing_files/#configmap-and-secrets-utility-functions
    files:
      - server.key
      - client.key
```

We would still need a separate mechanism to change the SSL_CLIENT_KEY_FILE etc variables, though.

Reply:
Would it be correct to say:

- `certificates` are loaded as CAs
- `keys` are client + server keys? (bit confusing to me here)

Reply:
certificates can be the certs for your keypairs too! like:

- ca-certs.crt -> bundle of certs
- server.crt -> certificate corresponding to the server cert (presented to users connecting to Coder)
- client.crt -> certificate corresponding to the client cert (presented to servers Coder is connecting to)

keys are the private keys corresponding to our certs (server or client), e.g.:

- server.key
- client.key

Reply:
Ahh I believe I understand.

I'm going to add these in another PR. I'm worried about scope-creep getting the best of us here.

Reply:
Yeah totally! I agree, a separate PR is better